### ISSUE TITLE
Should SA1118 (ParameterMustNotSpanMultipleLines) be reported for lambdas?

### ISSUE THREAD

**Issue by sharwell** [DotNetAnalyzers/StyleCopAnalyzers#820]
Is SA1118 supposed to be reported for the following statement?

```csharp
Assert.Equal(0, this.GetListQuery(ImmutableList<int>.Empty).FindLast(n =>
{
    Assert.True(false, "Predicate should not have been invoked.");
    return true;
}));
```

**Comment by vweijsters** [comment]
No, I don't think so. The documentation of the rule explicitly allows anonymous methods (and therefore lambda's) to span across multiple lines. See the quote below

>To prevent method calls from becoming excessively complicated and unreadable, only the first parameter to a method or indexer call is allowed to span across multiple lines. **The exception is an anonymous method passed as a parameter, which is always allowed to span multiple lines.** A violation of this rule occurs whenever a parameter other than the first parameter spans across multiple lines, and the parameter does not contain an anonymous method.

**Comment by sharwell** [comment]
Thanks, moving to a bug. :bug:

**Comment by macpak** [comment]
I will take.

**Comment by macpak** [comment]
Ok, SA1118 is not reported for:

    n =>
    {
        Assert.True(false, "Predicate should not have been invoked.");
        return true;
    }

but I guess it's reported for:

    this.GetListQuery(ImmutableList<int>.Empty).FindLast(n =>
    {
        Assert.True(false, "Predicate should not have been invoked.");
        return true;
     })

which is not lambda and spans multiple lines. I ran test:

       [Fact]
        public async Task TestMethodCallWithOneParameterWhichIsLambdaSpanningMultipleLines()
        {
            var testCode = @"
    class Foo
    {
        public void Fun(System.Func<int,bool> f)
        {
        }

        public void Bar()
        {
            Fun(n => 
            {
                var j = 5;
                return true;
            });
         }
    }";

            await this.VerifyCSharpDiagnosticAsync(testCode, EmptyDiagnosticResults, CancellationToken.None).ConfigureAwait(false);
        }

and it passes.

**Comment by alaendle** [comment]
btw, the rule should also exclude anonymous types I guess?
E.g.
```
        return this.RedirectToAction(
          "Error",
          new
            {
              returnUrl = Url.Action("Index", "AzureBrowsing", new { area = "Administration" }),
              returnUrlDisplay = Resources.Strings.strReturnToAzureBrowsingIndex
            });
```

**Comment by sharwell** [comment]
I'm not able to reproduce this issue, and the tests appear to already cover this behavior in a number of situations.

**Comment by alaendle** [comment]
At least I can confirm that the false positive does no longer occurs with beta 11. Thx. :+1:
